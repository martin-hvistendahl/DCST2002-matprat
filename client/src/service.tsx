import axios from 'axios';

axios.defaults.baseURL = 'http://localhost:3000/api/v2';

export type Recipe = {
  oppskrift_id: number;
  oppskrift_navn: string;
  oppskrift_beskrivelse: string;
  oppskrift_steg: string;
  ant_pors: number;
  bilde_adr: string;
  kategori_id: number;
  land_id: number;
  ant_like: number;
};

export type Recipe_Content = {
  oppskrift_id: number;
  ingred_id: number;
  mengde: number;
  maleenhet: string;
};

export type Country = {
  land_id: number;
  land_navn: string;
};
export type Category = {
  kategori_id: number;
  kategori_navn: string;
};
export type Ingredient = {
  ingred_id: number;
  ingred_navn: string;
};
class Service {
  /**
   * Get all tasks.
   */
  getAllRepice() {
    return axios.get<Recipe[]>('/').then((response) => response.data);
  }
  getRecipe(id: number) {
    return axios.get<Recipe[]>('/recipe/' + id).then((response) => response.data);
  }
  getRecipeContent(id: number) {
    return axios.get<Recipe_Content[]>('/recipecontent/' + id).then((response) => response.data);
  }
  getAllCountry() {
    return axios.get<Country[]>('/country').then((response) => response.data);
  }
  getAllCategory() {
    return axios.get<Category[]>('/category').then((response) => response.data);
  }
  getAllIngredient() {
    return axios.get<Ingredient[]>('/ingredient').then((response) => response.data);
  }
  createIngredient(name: string) {
    return axios.post<{}>('/newingredient', { name: name }).then((response) => response.data);
  }
  createCountry(name: string) {
    return axios.post<{}>('/newcountry', { name: name }).then((response) => response.data);
  }

  createRecipe(recipe: Recipe) {
    console.log(recipe);
    return axios
      .post<{ id: number }>('/createrecipe', { recipe: recipe })
      .then((response) => response.data.id);
  }
  createRecipeIngredient(recipe_content: Recipe_Content[]) {
    console.log('egentlig andre console lgo', recipe_content);
    return axios
      .post<Recipe_Content>('/create_recipe_ingredient', { recipe_content: recipe_content })
      .then((response) => response.data);
  }
  updateRecipeIngredient(recipeContent: Recipe_Content[]) {
    console.log('her motas dets', recipeContent);
    return axios
      .put('/update_recipe_ingredient', { recipeContent: recipeContent })
      .then((response) => response.data);
  }
  updateRecipe(recipe: Recipe) {
    return axios.put('/update_recipe', { recipe: recipe }).then((response) => response.data);
  }
}

const service = new Service();
export default service;
