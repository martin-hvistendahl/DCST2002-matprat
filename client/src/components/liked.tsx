import * as React from 'react';
import { Component } from 'react-simplified';
import { Alert, Cards, Rows, RecipeView } from '.././widgets';
import { NavLink } from 'react-router-dom';
import service, { Recipe } from '.././service';
import { createHashHistory } from 'history';

const history = createHashHistory(); // Use history.push(...) to programmatically change path, for instance after successfully saving a student
export class LikedRecipes extends Component {
  recipes: Recipe[] = [];

  render() {
    return (
      <div className="margintop">
        <div className="container-fluid">
          <Rows>{/* Filtrerer recipes etter hvilke som er likt og viser de med et card som linker til oppskriften */}
            {this.recipes
              .filter((recipe) => recipe.liked == true)
              .map((likedRecipe) => (
                <Cards title="" key={likedRecipe.oppskrift_id}>
                  <NavLink className="black" to={'/recipe/' + likedRecipe.oppskrift_id}>
                    <RecipeView
                      img={likedRecipe.bilde_adr}
                      name={likedRecipe.oppskrift_navn}
                      numbOfPors={likedRecipe.ant_pors}
                    ></RecipeView>
                  </NavLink>
                </Cards>
              ))}
          </Rows>
        </div>
      </div>
    );
  }

  mounted() {
    /* Henter alle oppskrftene */
    service
      .getAllRepice()
      .then((recipes) => {
        this.recipes = recipes;
      })
      .catch((error) => Alert.danger('Error getting tasks: ' + error.message));
  }
}
