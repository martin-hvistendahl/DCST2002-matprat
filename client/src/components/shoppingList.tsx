import * as React from 'react';
import { Component } from 'react-simplified';
import { Alert, Card, Row, Column, Form, Button, RecipeView } from '.././widgets';
import { NavLink, Redirect } from 'react-router-dom';
import service, {
  Country,
  Category,
  Ingredient,
  Recipe,
  Recipe_Content,
  List,
  ElementShoppingList,
} from '.././service';
import { createHashHistory } from 'history';

const history = createHashHistory(); // Use history.push(...) to programmatically change path, for instance after successfully saving a student
export class ShoppingList extends Component {
  shoppingList: List[] = [];
  ingredients: Ingredient[] = [];
  selectedIngredients: Ingredient[] = [];
  searchterm: string = '';
  elementHandleliste: ElementShoppingList = {
    ingred_id: 0,
    ingred_navn: '',
    mengde: 0,
    maleenhet: '',
  };
  selectedIngredient: List = {
    id: 0,
    ingred_id: 1,
    mengde: 0,
    maleenhet: '',
  };

  render() {
    return (
      <>
        <Card title="Handleliste">
          <Column>
            <div id="liste" className="">
              {this.shoppingList.map((sl, i) => (
                <Row key={i}>
                  <p style={{ width: '190px' }}>
                    {i + 1}.{' '}
                    {
                      this.ingredients.find((ingredient) => ingredient.ingred_id == sl.ingred_id)
                        ?.ingred_navn
                    }{' '}
                    {/* @ts-ignore */}
                  </p>
                  <input
                    className="form-control"
                    type="number"
                    style={{ width: '75px', marginRight: '10px' }}
                    onChange={(event) => {
                      //@ts-ignore
                      sl.mengde = event.currentTarget.value;
                      console.log(sl.mengde);
                    }}
                    onBlur={() => this.updatePortions(sl)}
                    value={sl.mengde}
                    size={2}
                  ></input>{' '}
                  <p style={{ width: '110px' }}>{sl.maleenhet}</p>
                  <Column width={1}>
                    <Button.Danger onClick={() => this.decrementPortions(sl)}>-</Button.Danger>
                  </Column>
                  <Column width={1}>
                    <Button.Success onClick={() => this.incrementPortions(sl)}>+</Button.Success>
                  </Column>
                  <Column width={1}>
                    <Button.Danger onClick={() => this.deleteIngredient(sl.id)}>
                      <svg
                        xmlns="http://www.w3.org/2000/svg"
                        width="16"
                        height="16"
                        fill="currentColor"
                        className="bi bi-trash"
                        viewBox="0 0 16 16"
                      >
                        <path d="M5.5 5.5A.5.5 0 0 1 6 6v6a.5.5 0 0 1-1 0V6a.5.5 0 0 1 .5-.5zm2.5 0a.5.5 0 0 1 .5.5v6a.5.5 0 0 1-1 0V6a.5.5 0 0 1 .5-.5zm3 .5a.5.5 0 0 0-1 0v6a.5.5 0 0 0 1 0V6z" />
                        <path
                          fillRule="evenodd"
                          d="M14.5 3a1 1 0 0 1-1 1H13v9a2 2 0 0 1-2 2H5a2 2 0 0 1-2-2V4h-.5a1 1 0 0 1-1-1V2a1 1 0 0 1 1-1H6a1 1 0 0 1 1-1h2a1 1 0 0 1 1 1h3.5a1 1 0 0 1 1 1v1zM4.118 4 4 4.059V13a1 1 0 0 0 1 1h6a1 1 0 0 0 1-1V4.059L11.882 4H4.118zM2.5 3V2h11v1h-11z"
                        />
                      </svg>
                    </Button.Danger>
                  </Column>
                </Row>
              ))}
              <Button.Danger onClick={() => this.deleteAll()}>Slett alle</Button.Danger>
            </div>
          </Column>
        </Card>
        <Card title="Legg til ingredienser">
          <Column>
            <p key={1}>
              Navn: {/* @ts-ignore */}
              <Form.Input
                id="navn"
                type="text"
                onChange={(event) => {
                  this.elementHandleliste.ingred_navn = event.currentTarget.value;
                }}
                value={this.elementHandleliste.ingred_navn}
              ></Form.Input>
              Antall: {/* @ts-ignore */}
              <Form.Input
                id="mengde"
                type="number"
                onChange={(event) => {
                  //@ts-ignore
                  this.elementHandleliste.mengde = event.currentTarget.value;
                }}
                value={this.elementHandleliste.mengde}
              ></Form.Input>
              Måleenhet: {/* @ts-ignore */}
              <Form.Input
                id="maleenhet"
                type="text"
                onChange={(event) => {
                  this.elementHandleliste.maleenhet = event.currentTarget.value;
                }}
                value={this.elementHandleliste.maleenhet}
              ></Form.Input>
              <Button.Success onClick={() => this.addItem(this.elementHandleliste)}>
                Legg til
              </Button.Success>
            </p>
          </Column>
        </Card>
        <Card title="Add existing ingredients">
          <Column>
            Søk:
            <Form.Input
              id="shoppinglistsearch"
              type="text"
              value={this.searchterm}
              onChange={(event) => {
                this.search(event.currentTarget.value);
                this.searchterm = event.currentTarget.value;
              }}
            />
            <select
              className="form-select"
              id="selectExistingIngredient"
              onChange={(event) => {
                this.selectedIngredient.ingred_id = Number(event.currentTarget.value);
              }}
            >
              {this.selectedIngredients.map((ingredient) => (
                <option key={ingredient.ingred_id} value={ingredient.ingred_id}>
                  {ingredient.ingred_navn}
                </option>
              ))}
            </select>
          
            Antall:{' '}
            <Form.Input
              id="exisitingmengde"
              type="number"
              value={this.selectedIngredient.mengde}
              onChange={(event) => {
                //@ts-ignore
                this.selectedIngredient.mengde = event.currentTarget.value;
              }}
            />
         
            Måleenhet:{' '}
            <Form.Input
              id="exisitingmaleenhet"
              type="text"
              value={this.selectedIngredient.maleenhet}
              onChange={(event) => {
                this.selectedIngredient.maleenhet = event.currentTarget.value;
                console.log(this.selectedIngredient);
              }}
            />
            <Button.Success onClick={() => this.addExistingItem(this.selectedIngredient)}>
              Legg til
            </Button.Success>
          </Column>
        </Card>
      </>
    );
  }

  mounted() {
    this.elementHandleliste.ingred_id = 0;
    this.elementHandleliste.ingred_navn = '';
    this.elementHandleliste.mengde = 0;
    this.elementHandleliste.maleenhet = '';

    this.searchterm = '';
    this.selectedIngredient.mengde = 0;
    this.selectedIngredient.maleenhet = '';

    service
      .getShoppingList()
      .then((shoppingList: List[]) => (this.shoppingList = shoppingList))
      .catch((error: { message: string }) =>
        Alert.danger('Error getting shoppingList: ' + error.message)
      );

    service
      .getAllIngredient()
      .then(
        (ingredients) => (
          (this.ingredients = ingredients),
          (this.selectedIngredients = ingredients),
          (this.selectedIngredient.ingred_id = document.getElementById('selectExistingIngredient')
            ? //@ts-ignore
              document.getElementById('selectExistingIngredient').value
            : '')
        )
      )
      .catch((error) => Alert.danger('Error getting ingredients: ' + error.message));
  }
  search(searchterm: string) {
    this.selectedIngredients = this.ingredients.filter((ingredient) =>
      ingredient.ingred_navn.toLowerCase().includes(searchterm.toLowerCase())
    );
    if (this.selectedIngredients.length > 0) {
      this.selectedIngredient.ingred_id = this.selectedIngredients[0].ingred_id;
    } else {
      this.selectedIngredient.ingred_id = 0;
      Alert.danger('Ingen ingredienser funnet');
    }
    console.log(this.selectedIngredients);
    console.log(this.selectedIngredient);
  }
  incrementPortions(ingredient: List) {
    ingredient.mengde++;
    service
      .updateIngredientShoppingList(ingredient)
      .then(() => this.mounted())
      .catch((error) => Alert.danger('Error getting ingredients: ' + error.message));
  }
  decrementPortions(ingredient: List) {
    if (ingredient.mengde > 1) {
      ingredient.mengde--;
      service
        .updateIngredientShoppingList(ingredient)
        .then(() => this.mounted())
        .catch((error) => Alert.danger('Error getting ingredients: ' + error.message));
    }
  }
  updatePortions(ingredient: List) {
    //@ts-ignore
    if (
      ingredient.mengde < 0 ||
      ingredient.mengde == null ||
      ingredient.mengde == undefined ||
      //@ts-ignore
      ingredient.mengde == ''
    ) {
      ingredient.mengde = 1;
    }
    service
      .updateIngredientShoppingList(ingredient)
      .then(() => this.mounted())
      .then(() => Alert.info('Antall oppdatert'))
      .catch((error) => Alert.danger('Error getting ingredients: ' + error.message));
  }

  deleteIngredient(id: number) {
    service
      .deleteIngredientShoppingList(id)
      .then(() => this.mounted())
      .catch((error) => Alert.danger('Error getting ingredients: ' + error.message));
  }

  deleteAll() {
    service
      .deleteAllShoppingList()
      .then(() => this.mounted())
      .catch((error) => Alert.danger('Error deleting shopping list: ' + error.message));
  }

  addItem(item: ElementShoppingList) {
    console.log(this.ingredients.length);
    console.log(this.ingredients);
    item.ingred_id = this.ingredients[this.ingredients.length - 1].ingred_id + 1;
    if (item.ingred_navn == null || item.ingred_navn == undefined || item.ingred_navn == '') {
      Alert.danger('Du må fylle inn navn på ingrediensen');
    } else if (
      this.ingredients.some(
        (ing) => ing.ingred_navn.toLowerCase() == item.ingred_navn.toLowerCase()
      ) == true
    ) {
      Alert.danger('Ingrediensen eksisterer allerede');
    } else if (
      item.mengde == null ||
      item.mengde == undefined ||
      //@ts-ignore
      item.mengde == '' ||
      item.mengde < 0 ||
      item.mengde == 0
    ) {
      item.mengde = 1;
      Alert.danger('Du må fylle inn antall av ingrediensen, dette må være heltall større enn 0');
      return false;
    } else if (item.maleenhet == null || item.maleenhet == undefined || item.maleenhet == '') {
      Alert.danger('Du må fylle inn måleenhet');
      return false;
    } else {
      service
        .createIngredient(item.ingred_navn)
        .then(() => {
          service
            .addIngredient(item)
            .then(() => this.mounted())
            .catch((error) => Alert.danger('Error adding item to shopping list: ' + error.message));
        })
        .catch((error) => Alert.danger('Error creating new ingredient: ' + error.message));
    }
  }

  addExistingItem(item: List) {
    if (
      item.mengde == null ||
      item.mengde == undefined ||
      //@ts-ignore
      item.mengde == '' ||
      item.mengde < 0
    ) {
      item.mengde = 1;
      Alert.danger('Du må fylle inn antall av ingrediensen, dette må være heltall større enn 0');
    } else if (item.maleenhet == null || item.maleenhet == undefined || item.maleenhet == '') {
      item.maleenhet = 'stk';
      Alert.danger('Du må fylle inn måleenhet');
    } else if (
      item.ingred_id == 0 ||
      item.ingred_id < 0 ||
      item.ingred_id == null ||
      item.ingred_id == undefined
    ) {
      Alert.danger('Du må velge en ingrediens');
    } else {
      service
        .addIngredient(item)
        .then(() => this.mounted())
        .catch((error) => Alert.danger('Error adding item to shopping list: ' + error.message));
    }
  }
}
