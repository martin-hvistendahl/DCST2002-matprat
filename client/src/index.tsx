import ReactDOM from 'react-dom';
import * as React from 'react';
import { Component } from 'react-simplified';
import { NavLink, HashRouter, Route } from 'react-router-dom';
import { NavBar, Card, Alert, Column, Row, Form, Button, RecipeView } from './widgets';
import { NewRecipe } from './components';

import service, { Recipe } from './service';
import { createHashHistory } from 'history';

class Menu extends Component {
  render() {
    return (
      <NavBar brand="MatForum">
        <NavBar.Link to="/newrecipe">Ny oppskrift</NavBar.Link>
      </NavBar>
    );
  }
}

class Home extends Component {
  recipes: Recipe[] = [];

  render() {
    return (
      <>
        <Card title="Oppskrifter">
          {this.recipes.map((recipe) => (
            <Row key={recipe.oppskrift_id}>
              <Column>
                <NavLink
                  style={{ textDecoration: 'none', color: 'black' }}
                  to={'/recipe/' + recipe.oppskrift_id}
                >
                  <RecipeView
                    img={recipe.bilde_adr}
                    name={recipe.oppskrift_navn}
                    numbOfPors={recipe.ant_pors}
                  ></RecipeView>
                </NavLink>
              </Column>
            </Row>
          ))}
        </Card>
      </>
    );
  }

  mounted() {
    service
      .getAllRepice()
      .then((recipes) => (this.recipes = recipes))
      .catch((error) => Alert.danger('Error getting tasks: ' + error.message));
  }
}

ReactDOM.render(
  <HashRouter>
    <div>
      <Alert />
      <Menu />
      <Route exact path="/" component={Home} />
      <Route exact path="/newrecipe" component={NewRecipe} />
    </div>
  </HashRouter>,
  document.getElementById('root')
);
