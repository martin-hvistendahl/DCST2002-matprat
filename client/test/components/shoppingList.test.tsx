import * as React from 'react';
import userEvent from '@testing-library/user-event';
import { shallow, mount } from 'enzyme';
import { Alert, Form, Button } from '../../src/widgets';
import { ShoppingList } from '../../src/components/shoppingList';
import { List } from '../../src/service';

import { createHashHistory } from 'history';

const history = createHashHistory();
jest.mock('../../src/service', () => {
  class Service {
    getAllIngredient() {
      return Promise.resolve([
        {
          ingred_id: 1,
          ingred_navn: 'Chicken',
        },
        {
          ingred_id: 2,
          ingred_navn: 'Salmon',
        },
        {
          ingred_id: 3,
          ingred_navn: 'Beef',
        },
      ]);
    }
    getShoppingList() {
      return Promise.resolve([
        { id: 59, ingred_id: 1, mengde: 2, maleenhet: 'stk' },
        {
          id: 60,
          ingred_id: 2,
          mengde: 1,
          maleenhet: 'kg',
        },
      ]);
    }
    deleteAllShoppingList() {
      return Promise.resolve();
    }
    deleteShoppingList(id: number) {
      return Promise.resolve();
    }
    updateIngredientShoppingList(ingredient: List) {
      return Promise.resolve();
    }
    deleteIngredientShoppingList(id: number) {
      return Promise.resolve();
    }
    createIngredient(item: string) {
      return Promise.resolve();
    }
    addIngredient(item: any) {
      return Promise.resolve();
    }
  }
  return new Service();
});

describe('Rending', () => {
  test('Should render after snapshot', (done) => {
    const wrapper = shallow(<ShoppingList />);
    setTimeout(() => {
      expect(wrapper).toMatchSnapshot();

      done();
    });
  });
});
describe('Functionality input', () => {
  test('Change input field on ingrient in shoppinglist', (done) => {
    const wrapper = shallow(<ShoppingList />);

    setTimeout(() => {
      expect(wrapper.find(Form.Input).at(3).prop('value')).toBe(2);

      wrapper
        .find(Form.Input)
        .at(3)
        .simulate('change', { currentTarget: { value: '6.5' } });

      expect(wrapper.find(Form.Input).at(3).prop('value')).toBe('6.5');

      wrapper.find(Form.Input).at(3).simulate('onBlur');

      expect(wrapper.find(Form.Input).at(3).prop('value')).toEqual('6.5');

      done();
    });
  });
  test('Add new ingredient to shopping list', (done) => {
    const wrapper = mount(<ShoppingList />);
    setTimeout(() => {
      wrapper.find('#choseIngredient').at(0).simulate('keyDown', { key: 'ArrowDown', keyCode: 40 });
      wrapper.find('#choseIngredient').at(0).simulate('keyDown', { key: 'ArrowDown', keyCode: 40 });
      wrapper.find('#choseIngredient').at(0).simulate('keyDown', { key: 'Enter', keyCode: 13 });

      wrapper
        .find('#exisitingmengde')
        .at(0)
        .simulate('change', { currentTarget: { value: 3 } });
      wrapper
        .find('#exisitingmaleenhet')
        .at(0)
        .simulate('change', { currentTarget: { value: 'plater' } });

      wrapper.find(Button.Success).at(0).simulate('click');

      setTimeout(() => {
        //får ikke testet om det er lagt til i listen siden det er et database kall
        //men kan sjekke om input feltete blir nullet ut, dette vil indikere at det er lagt til i listen
        expect(wrapper.find('#choseIngredient').at(0).first().text()).toEqual('Chicken');
        expect(wrapper.find('#exisitingmengde').at(0).prop('value')).toBe('0');
        expect(wrapper.find('#exisitingmaleenhet').at(0).prop('value')).toBe('');
        //expect wrapper to mount
        expect(wrapper.find(Button.Success).at(0).simulate('click')).toBeTruthy();
        done();
      });
    });
  });

  test('No ingredient in inputfield when creating new ingredient', (done) => {
    const wrapper = shallow(<ShoppingList />);
    const wrapperAlert = shallow(<Alert />);
    setTimeout(() => {
      wrapper.find('#createIngredient').simulate('change', { currentTarget: { value: '' } });
      wrapper.find(Button.Success).at(1).simulate('click');

      setTimeout(() => {
        expect(
          wrapperAlert.containsMatchingElement(
            <div>
              Du må fylle inn navn på ingrediensen<button></button>
            </div>
          )
        ).toEqual(true);
        done();
      });
    });
  });
});

describe('Functionality buttons', () => {
  test('Delete all items', (done) => {
    const wrapper = shallow(<ShoppingList />);
    setTimeout(() => {
      wrapper.find(Button.Danger).at(4).simulate('click');

      setTimeout(() => {
        expect(wrapper.find(Button.Danger).at(4)).toEqual({});
        done();
      });
    });
  });
  test('No ingredient, click add button get error', (done) => {
    const wrapper = shallow(<ShoppingList />);
    const wrapperAlert = shallow(<Alert />);
    setTimeout(() => {
      wrapper.find('#exisitingmengde').simulate('change', { currentTarget: { value: 2 } });
      wrapper.find(Button.Success).at(0).simulate('click');

      setTimeout(() => {
        expect(
          wrapperAlert.containsMatchingElement([
            <div>
              Du må fylle inn måleenhet<button></button>
            </div>,
          ])
        );
        // expect(wrapper.find(Button.Success).at(1)).toEqual({});
        done();
      });
    });
  });
  test('Create a new existing ingridient', (done) => {
    const wrapper = shallow(<ShoppingList />);
    const wrapperAlert = shallow(<Alert />);
    setTimeout(() => {
      wrapper.find('#createIngredient').simulate('change', { currentTarget: { value: 'Chicken' } });
      wrapper.find(Button.Success).at(1).simulate('click');

      setTimeout(() => {
        expect(
          wrapperAlert.containsMatchingElement([
            <div>
              Ingrediensen eksisterer allerede<button></button>
            </div>,
          ])
        );

        done();
      });
    });
  });
  test('Create a new ingredient', (done) => {
    const wrapper = shallow(<ShoppingList />);
    setTimeout(() => {
      wrapper.find('#createIngredient').simulate('change', { currentTarget: { value: 'Melk' } });
      wrapper.find(Button.Success).at(1).simulate('click');
    });

    setTimeout(() => {
      expect(wrapper.find(Button.Success).at(1)).toEqual({});

      done();
    });
  });

  test('Increment mengde', (done) => {
    const wrapper = shallow(<ShoppingList />);
    setTimeout(() => {
      // må ha en timeout her fordi den ikke klarer å finne knappen uten
      expect(wrapper.find(Form.Input).at(3).prop('value')).toBe(2);

      wrapper.find(Button.Success).at(2).simulate('click');

      expect(wrapper.find(Form.Input).at(3).prop('value')).toBe('3');

      expect(wrapper.find(Form.Input).at(4).prop('value')).toBe(1);

      wrapper.find(Button.Success).at(3).simulate('click');

      expect(wrapper.find(Form.Input).at(4).prop('value')).toBe('2');

      done();
    });
  });

  test('Decrement mengde', (done) => {
    const wrapper = shallow(<ShoppingList />);
    setTimeout(() => {
      // må ha en timeout her fordi den ikke klarer å finne knappen uten
      expect(wrapper.find(Form.Input).at(3).prop('value')).toBe(2);

      wrapper.find(Button.Danger).at(0).simulate('click');

      expect(wrapper.find(Form.Input).at(3).prop('value')).toBe('1');

      expect(wrapper.find(Form.Input).at(4).prop('value')).toBe(1);

      wrapper.find(Button.Danger).at(2).simulate('click');

      expect(wrapper.find(Form.Input).at(4).prop('value')).toBe(1);

      done();
    });
  });

  test('Delete item', (done) => {
    const wrapper = shallow(<ShoppingList />);
    setTimeout(() => {
      // må ha en timeout her fordi den ikke klarer å finne knappen uten
      wrapper.find(Button.Danger).at(1).simulate('click');

      setTimeout(() => {
        expect(wrapper.find(Button.Danger).at(1)).toEqual({});
        done();
      });
    });
  });
  test('Onblur trigger', (done) => {
    const wrapper = shallow(<ShoppingList />);
    setTimeout(() => {
      wrapper
        .find(Form.Input)
        .at(3)
        .simulate('blur', { currentTarget: { value: 2 } });
      setTimeout(() => {
        userEvent.tab();

        setTimeout(() => {
          expect(wrapper.find(Form.Input).at(3).prop('value')).toBe(2);
          done();
        });
      });
    });
  });
});
