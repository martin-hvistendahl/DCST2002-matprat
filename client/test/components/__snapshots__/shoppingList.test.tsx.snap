// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rending Should render after snapshot 1`] = `
<Fragment>
  <Card
    title="Handleliste"
  >
    <Column>
      <p
        key="0"
      >
        1
        .
         
        pizzadeig
         
        <input
          onBlur={[Function]}
          onChange={[Function]}
          size={2}
          type="string"
          value={2}
        />
         
        stk
        <ButtonDanger
          onClick={[Function]}
        >
          -
        </ButtonDanger>
        <ButtonSuccess
          onClick={[Function]}
        >
          +
        </ButtonSuccess>
        <ButtonDanger
          onClick={[Function]}
        >
          <svg
            className="bi bi-trash"
            fill="currentColor"
            height="16"
            viewBox="0 0 16 16"
            width="16"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M5.5 5.5A.5.5 0 0 1 6 6v6a.5.5 0 0 1-1 0V6a.5.5 0 0 1 .5-.5zm2.5 0a.5.5 0 0 1 .5.5v6a.5.5 0 0 1-1 0V6a.5.5 0 0 1 .5-.5zm3 .5a.5.5 0 0 0-1 0v6a.5.5 0 0 0 1 0V6z"
            />
            <path
              d="M14.5 3a1 1 0 0 1-1 1H13v9a2 2 0 0 1-2 2H5a2 2 0 0 1-2-2V4h-.5a1 1 0 0 1-1-1V2a1 1 0 0 1 1-1H6a1 1 0 0 1 1-1h2a1 1 0 0 1 1 1h3.5a1 1 0 0 1 1 1v1zM4.118 4 4 4.059V13a1 1 0 0 0 1 1h6a1 1 0 0 0 1-1V4.059L11.882 4H4.118zM2.5 3V2h11v1h-11z"
              fillRule="evenodd"
            />
          </svg>
        </ButtonDanger>
      </p>
      <p
        key="1"
      >
        2
        .
         
        pizza fyll
         
        <input
          onBlur={[Function]}
          onChange={[Function]}
          size={2}
          type="string"
          value={1}
        />
         
        L
        <ButtonDanger
          onClick={[Function]}
        >
          -
        </ButtonDanger>
        <ButtonSuccess
          onClick={[Function]}
        >
          +
        </ButtonSuccess>
        <ButtonDanger
          onClick={[Function]}
        >
          <svg
            className="bi bi-trash"
            fill="currentColor"
            height="16"
            viewBox="0 0 16 16"
            width="16"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M5.5 5.5A.5.5 0 0 1 6 6v6a.5.5 0 0 1-1 0V6a.5.5 0 0 1 .5-.5zm2.5 0a.5.5 0 0 1 .5.5v6a.5.5 0 0 1-1 0V6a.5.5 0 0 1 .5-.5zm3 .5a.5.5 0 0 0-1 0v6a.5.5 0 0 0 1 0V6z"
            />
            <path
              d="M14.5 3a1 1 0 0 1-1 1H13v9a2 2 0 0 1-2 2H5a2 2 0 0 1-2-2V4h-.5a1 1 0 0 1-1-1V2a1 1 0 0 1 1-1H6a1 1 0 0 1 1-1h2a1 1 0 0 1 1 1h3.5a1 1 0 0 1 1 1v1zM4.118 4 4 4.059V13a1 1 0 0 0 1 1h6a1 1 0 0 0 1-1V4.059L11.882 4H4.118zM2.5 3V2h11v1h-11z"
              fillRule="evenodd"
            />
          </svg>
        </ButtonDanger>
      </p>
      <ButtonDanger
        onClick={[Function]}
      >
        Slett alle
      </ButtonDanger>
    </Column>
  </Card>
  <Card
    title="Legg til ingredienser"
  >
    <Column>
      <p
        key="1"
      >
        Navn:
         
        <input
          id="navn"
          onChange={[Function]}
          type="text"
          value=""
        />
        Antall:
         
        <input
          id="mengde"
          onChange={[Function]}
          type="number"
          value={0}
        />
        Måleenhet:
         
        <input
          id="maleenhet"
          onChange={[Function]}
          type="text"
          value=""
        />
        <ButtonSuccess
          onClick={[Function]}
        >
          Legg til
        </ButtonSuccess>
      </p>
    </Column>
  </Card>
  <Card
    title="Add existing ingredients"
  >
    <Column>
      <h6>
        Søk
      </h6>
      <FormInput
        id="shoppinglistsearch"
        onChange={[Function]}
        type="text"
        value=""
      />
      <select
        id="selectExistingIngredient"
        onChange={[Function]}
      >
        <option
          key="1"
          value={1}
        >
          pizzadeig
        </option>
        <option
          key="2"
          value={2}
        >
          pizza fyll
        </option>
        <option
          key="3"
          value={3}
        >
          kjøttboller
        </option>
      </select>
    </Column>
    <Column>
      Antall:
       
      <FormInput
        id="exisitingmengde"
        onChange={[Function]}
        type="number"
        value={0}
      />
    </Column>
    <Column>
      Måleenhet:
       
      <FormInput
        id="exisitingmaleenhet"
        onChange={[Function]}
        type="text"
        value=""
      />
    </Column>
    <Column>
      <ButtonSuccess
        onClick={[Function]}
      >
        Legg til
      </ButtonSuccess>
    </Column>
  </Card>
</Fragment>
`;
